src/output.js

```diff
@@ -3,36 +3,90 @@ const path = require("path");
 const { config } = require("./config");
 
 function saveToCSV(filename, data) {
-  const headers = Object.keys(data[0]);
-  const filteredHeaders = headers.filter((header) => !header.startsWith("_"));
-  const headerRow = filteredHeaders.join(",") + "\n";
-  const dataRows = data
-    .map((row) => filteredHeaders.map((header) => row[header]).join(","))
-    .join("\n");
-  const csv = headerRow + dataRows;
-  writeFile(filename, csv);
+  return new Promise((resolve, reject) => {
+    if (!data || data.length === 0) {
+      console.error(`No data provided to write to ${filename}`);
+      return reject(new Error("Data is empty."));
+    }
+
+    const headers = Object.keys(data[0]);
+    const filteredHeaders = headers.filter((header) => !header.startsWith("_"));
+
+    // Create a write stream
+    const stream = fs.createWriteStream(filename);
+
+    // Handle stream errors and log them
+    stream.on('error', (err) => {
+      console.error(`Error writing to CSV file ${filename}: ${err.message}`);
+      reject(err);
+    });
+
+    // Write the headers first
+    stream.write(filteredHeaders.join(",") + "\n");
+
+    // Write data rows incrementally
+    data.forEach((row) => {
+      const rowData = filteredHeaders.map((header) => row[header] || "").join(",");
+      stream.write(rowData + "\n");
+    });
```

Comment:
We can improve the performance by batching the rows.

```
const batchSize = 10000;
    let batch = [];

    data.forEach((row) => {
      const rowData = filteredHeaders.map((header) => row[header] || "").join(",");
      batch.push(rowData);

      if (batch.length === batchSize) {
        stream.write(batch.join("\n") + "\n");
        batch = [];
      }
    });

    //Write any remaining rows
    if (batch.length > 0) {
      stream.write(batch.join("\n") + "\n");
    }

Reply:
This is great, I love it. Do you prefer making batchSize configurable and define in config.js as `BATCH_SIZE_STREAM_TO_FILE: 10000` or simply hardcode it as 10000 (I think 10k is a safe and conservative batch size)

Reply:
I think it's fine for now to hardcode a batch size. We know that 10k works, so it should be safe.

Reply:
Sounds good, thanks!